refactor(Analysis/SpecialFunctions): extend `tendsto_one_plus_div_rpow_exp`

hanwenzhu opened the pull request:
This PR

- Proves 12 theorems (they are stated separately because of differences in `rexp` vs `cexp` and `pow` vs `rpow` vs `cpow`):
    - The limit of `x * log (1 + t/x + o(1/x))` as `x → ∞` is `t` for `t : ℂ` or `ℝ` and `x : ℝ` or `ℕ`.
    - The limit of `(1 + t/x + o(1/x)) ^ x` as `x → ∞` is `t` for `t : ℂ` or `ℝ` and `x : ℝ` or `ℕ`.
    - The limit of `(1 + t/x) ^ x` as `x → ∞` is `t` for `t : ℂ` or `ℝ` and `x : ℝ` or `ℕ`.
    - (Note: the last row of theorems for `t : ℝ` already exist and are now replaced)
- Renames `plus` to `add` in affected theorems

The little-o theorem for `t : ℂ` is needed for [CLT](https://github.com/RemyDegenne/CLT).

Moves:
- tendsto_one_plus_div_rpow_exp -> Real.tendsto_one_add_div_rpow_exp
- tendsto_one_plus_div_pow_exp -> Real.tendsto_one_add_div_pow_exp

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

j-loreaux commented on Mathlib/Analysis/SpecialFunctions/Complex/LogBounds.lean:
```diff
@@ -267,3 +267,180 @@ lemma hasSum_taylorSeries_neg_log {z : ℂ} (hz : ‖z‖ < 1) :
+private lemma tendsto_g_comp_floor {g : ℕ → ℂ} {t : ℂ}
+    (hg : Tendsto (fun n ↦ n * g n) atTop (𝓝 t)) :
+    Tendsto (fun x : ℝ ↦ x * g ⌊x⌋₊) atTop (𝓝 t) := by
+  -- for these two lemmas, we copy the proof over ℝ from `g_le_div` and `g_lt`
```
This is the key thing that should inform you that you need to generalize your earlier results. Or alternatively, I think after 
```
apply (hg.comp tendsto_nat_floor_atTop).congr_dist
```
you can just use [Filter.isBoundedUnder_le_mul_tendsto_zero](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Normed/Ring/Lemmas.html#Filter.isBoundedUnder_le_mul_tendsto_zero). This should also also you to subsume the next lemma too with this one, if I'm not mistaken.

hanwenzhu replied:
Thanks for the suggestion, I used this theorem instead.

j-loreaux reviewed (commented):
These are some nice results, but I strongly encourage refactoring the `private` lemmas you have at the beginning of the code. You need to generalize and make them more reusable. If you get stuck, ping me and I can provide more explicit direction.

hanwenzhu commented on Mathlib/Analysis/SpecialFunctions/Complex/LogBounds.lean:
```diff
@@ -267,3 +267,180 @@ lemma hasSum_taylorSeries_neg_log {z : ℂ} (hz : ‖z‖ < 1) :
+open scoped Topology
+
+private lemma g_le_div {g : ℝ → ℂ} {t : ℂ} (hg : Tendsto (fun x ↦ x * g x) atTop (𝓝 t)) :
+    ∀ᶠ x in atTop, ‖g x‖ ≤ (‖t‖ + 1) / |x| := by
```
I don't really need the constant. This private lemma is refactored away.

hanwenzhu commented on Mathlib/Analysis/SpecialFunctions/Complex/LogBounds.lean:
```diff
@@ -267,3 +267,180 @@ lemma hasSum_taylorSeries_neg_log {z : ℂ} (hz : ‖z‖ < 1) :
+  simpa using h
+
+private lemma g_lt {g : ℝ → ℂ} {t : ℂ} (hg : Tendsto (fun x ↦ x * g x) atTop (𝓝 t))
+    (y : ℝ) (hy : 0 < y := by norm_num) : ∀ᶠ x in atTop, ‖g x‖ < y := by
```
Thanks for the suggestion! I proved this in `tendsto_zero_of_isBoundedUnder_mul_of_tendsto_cobounded`.

hanwenzhu commented:
I have some questions:
1. I implemented `RCLike.tendsto_mul_comp_nat_floor_of_tendsto_mul` (it is RCLike because I need it for both R and C, and it holds for both). However, the current layout makes the imports much larger because this theorem in `Mathlib.Analysis.SpecificLimits.RCLike` also depends on something I added to `Mathlib.Analysis.SpecificLimits.Normed`. I don't know the best layout of these theorems.
2. This `RCLike.tendsto_mul_comp_nat_floor_of_tendsto_mul` theorem is likely generalizable to more general rings in which `Tendsto Nat.cast atTop (cobounded R)`, but I don't know in which rings this holds. `RCLike` is just the most general typeclass I could find, for which this holds.
3. Should the derived limits be in a new module under `Analysis.SpecificLimits`?

dupuisf commented on Mathlib/Analysis/SpecialFunctions/Complex/LogBounds.lean:
```diff
@@ -267,3 +287,132 @@ lemma hasSum_taylorSeries_neg_log {z : ℂ} (hz : ‖z‖ < 1) :
+
+namespace Complex
+
+/-- The limit of `x * log (1 + t/x + o(1/x))` as `(x : ℝ) → ∞` is `t` for `t ∈ ℂ`. -/
+lemma tendsto_mul_log_one_add_of_tendsto {g : ℝ → ℂ} {t : ℂ}
+    (hg : Tendsto (fun x ↦ x * g x) atTop (𝓝 t)) :
+    Tendsto (fun x ↦ x * log (1 + g x)) atTop (𝓝 t) := by
```
Can you fix the docstrings to match the lemma statements here and below?

hanwenzhu replied:
Do you mean something like this? 
```
The limit of `x * log (1 + g x)` as `(x : ℝ) → ∞` equals `t : ℂ`, where `t` is the limit of `x * g(x)`.
```

I am happy to change it in this case. The reason I am using the little-o notation in the docstring is for potentially easier search (I don't know if this makes sense), because that's the notation commonly used in informal language.

dupuisf replied:
Yes, something like that. But if you want a statement that matches the natural language version better, you could probably go with something like
```lean
/-- The limit of `x * log (1 + t/x + o(1/x))` as `(x : ℝ) → ∞` is `t` for `t ∈ ℂ`. -/
lemma tendsto_mul_log_one_add_of_tendsto {g : ℝ → ℂ} {t : ℂ}
    (hg : g =o[atTop] fun x => 1 / x) :
    Tendsto (fun x ↦ x * log (1 + t / x + g x)) atTop (𝓝 t) := by sorry
```

dupuisf commented:
>     2. This `RCLike.tendsto_mul_comp_nat_floor_of_tendsto_mul` theorem is generalizable to more general rings in which `Tendsto Nat.cast atTop (cobounded R)`, but I don't know in which rings this holds. `RCLike` is just the most general typeclass I could find, for which this holds.

I guess this should be true for any ring where `[NormSMulClass R]` holds, no?

hanwenzhu commented:
> > ```
> > 2. This `RCLike.tendsto_mul_comp_nat_floor_of_tendsto_mul` theorem is generalizable to more general rings in which `Tendsto Nat.cast atTop (cobounded R)`, but I don't know in which rings this holds. `RCLike` is just the most general typeclass I could find, for which this holds.
> > ```
> 
> I guess this should be true for any ring where `[NormSMulClass R]` holds, no?

Thanks! I think that's true. I will put on awaiting-author and figure this out first.

hanwenzhu commented:
When I was writing the code for determining the general typeclasses for which `Tendsto Nat.cast atTop (cobounded R)` holds, I noticed one can prove the following:

```lean
-- in this PR
instance [RCLike K] : NormSMulClass ℤ K
```

and once there are instances of `NormSMulClass ℤ` then the following are useful

```lean
-- not in this PR, has name conflict with `norm_zsmul`
theorem norm_zsmul' {α} [SeminormedAddCommGroup α] [NormSMulClass ℤ α] (n : ℤ) (x : α) :
    ‖n • x‖ = |n| * ‖x‖ := by
  rw [norm_smul, Int.norm_eq_abs, Int.cast_abs]

-- not in this PR
theorem norm_nsmul {α} [SeminormedAddCommGroup α] [NormSMulClass ℤ α] (n : ℕ) (x : α) :
    ‖n • x‖ = n * ‖x‖ := by
  simpa using norm_zsmul' n x

-- in this PR
theorem norm_intCast_eq_abs_mul_norm_one (α) [SeminormedRing α] [NormSMulClass ℤ α] (n : ℤ) :
    ‖(n : α)‖ = |n| * ‖(1 : α)‖ := by
  rw [← zsmul_one, norm_zsmul']

-- in this PR
theorem norm_natCast_eq_abs_mul_norm_one (α) [SeminormedRing α] [NormSMulClass ℤ α] (n : ℕ) :
    ‖(n : α)‖ = n * ‖(1 : α)‖ := by
  simpa using norm_intCast_eq_abs_mul_norm_one α n

@[simp]
theorem norm_intCast (α) [SeminormedRing α] [NormOneClass α] [NormSMulClass ℤ α] (n : ℤ) :
    ‖(n : α)‖ = |n| := by
  simp [norm_intCast_eq_abs_mul_norm_one]

@[simp]
theorem norm_natCast (α) [SeminormedRing α] [NormOneClass α] [NormSMulClass ℤ α] (n : ℕ) :
    ‖(n : α)‖ = n := by
  simp [norm_natCast_eq_abs_mul_norm_one]
```

With the correct instances the last two in particular are generalizations of `Complex.norm_intCast` and `(Int|Real|Complex|RCLike).norm_natCast`; the first is a variation of current `norm_zsmul`. I am not sure about the first two because they just look like cheap corollaries of `NormSMulClass ℤ α`. The last two are not included, because they cause name conflict with e.g. `Real.norm_natCast`, and they are not needed for the proof of this PR. I think including them is good for a future PR, but would require a refactor in many places, and would require thinking about generalization to `nnnorm`, `enorm`, `ratCast`, `nnratCast`.

dupuisf commented:
Looks good, thanks!

bors r+
